### PR TITLE
Shared `View CFG` implementation

### PR DESCRIPTION
**Pull request by hvitved**
This PR moves the existing `View CFG` implementations for Ruby and Swift into the shared CFG library, and subsequently instantiates the shared implementation in C#.

### CODE REVIEW

File: shared/controlflow/codeql/controlflow/Cfg.qll [hunk 1]
```diff
@@ -1192,6 +1193,78 @@ module Make<LocationSig Location, InputSig<Location> Input> {
...
+  /** Provides the input to `ViewCfgQuery`. */
+  signature module ViewCfgQueryInputSig<FileSig File> {
+    /** The source file selected in the IDE. Should be an `external` predicate. */
+    string selectedSourceFile();
+
+    /** The source line selected in the IDE. Should be an `external` predicate. */
+    string selectedSourceLine();
+
+    /** The source column selected in the IDE. Should be an `external` predicate. */
+    string selectedSourceColumn();
```
Comment (aibaars) [inline shared/controlflow/codeql/controlflow/Cfg.qll:1205]: A relatively recent version of CodeQL should be able to deal with non-string types for external predicates. In any case we should have a sensible type for lines and column numbers in the signature module.

```suggestion
    int selectedSourceLine();

    /** The source column selected in the IDE. Should be an `external` predicate. */
    int selectedSourceColumn();
```

File: shared/controlflow/codeql/controlflow/Cfg.qll [hunk 2]
```diff
@@ -1192,6 +1193,78 @@ module Make<LocationSig Location, InputSig<Location> Input> {
     }
   }
 
+  /** Provides the input to `ViewCfgQuery`. */
+  signature module ViewCfgQueryInputSig<FileSig File> {
+    /** The source file selected in the IDE. Should be an `external` predicate. */
+    string selectedSourceFile();
+
+    /** The source line selected in the IDE. Should be an `external` predicate. */
+    string selectedSourceLine();
```
Comment (aibaars) [inline shared/controlflow/codeql/controlflow/Cfg.qll:1202]: Question: is this predicate really needed in the signature. Can't we simply define `external` predicates in the shared library?

File: shared/controlflow/codeql/controlflow/Cfg.qll (hunk 1 above)
Comment (hvitved) [reply shared/controlflow/codeql/controlflow/Cfg.qll:1205]: Not sure; then all languages would have to make the `.toInt` calls.

File: shared/controlflow/codeql/controlflow/Cfg.qll (hunk 2 above)
Comment (hvitved) [reply shared/controlflow/codeql/controlflow/Cfg.qll:1202]: External predicates can't be defined in (parameterized) modules.

File: swift/ql/lib/codeql/swift/controlflow/internal/PrintCFG.ql [hunk 3]
```diff
@@ -7,52 +7,36 @@
...
+private import codeql.swift.elements.File
 private import codeql.swift.controlflow.ControlFlowGraph
-private import codeql.swift.controlflow.internal.ControlFlowGraphImpl::TestOutput
+private import codeql.swift.controlflow.internal.ControlFlowGraphImpl as Impl
+private import codeql.swift.controlflow.internal.ControlFlowGraphImplSpecific
 
-/**
- * Gets the source file to generate a CFG from.
- */
 external string selectedSourceFile();
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/controlflow/internal/PrintCFG.ql:15]: Why have the comments on `selectedSourceFile`, `selectedSourceLine` and `selectedSourceColumn` been deleted?

**Review by geoffw0 — COMMENTED** [review]
Thanks for this.  Swift changes otherwise LGTM.  It's nice to have the explanations of how an file name extension etc are defined.

File: shared/controlflow/codeql/controlflow/Cfg.qll (hunk 1 above)
Comment (hvitved) [reply shared/controlflow/codeql/controlflow/Cfg.qll:1205]: Oh, I thought we would have to make changes to whatever logic constructs data for `selectedSourceLine` as well.

File: shared/controlflow/codeql/controlflow/Cfg.qll [hunk 4]
```diff
@@ -1192,6 +1193,78 @@ module Make<LocationSig Location, InputSig<Location> Input> {
     }
   }
 
+  /** Provides the input to `ViewCfgQuery`. */
+  signature module ViewCfgQueryInputSig<FileSig File> {
+    /** The source file selected in the IDE. Should be an `external` predicate. */
+    string selectedSourceFile();
```
Comment (aibaars) [inline shared/controlflow/codeql/controlflow/Cfg.qll:1199]: Can we move the file, line, and column predicates to the `IdeContextual` module?

**Review by michaelnebel — APPROVED** [review]
Uh, I am looking forward to trying it out!

File: shared/controlflow/codeql/controlflow/Cfg.qll (hunk 4 above)
Comment (hvitved) [reply shared/controlflow/codeql/controlflow/Cfg.qll:1199]: Not sure what you mean; we can't move the `external` predicates, as they cannot be defined inside modules.

**Review by aibaars — APPROVED** [review]

**Pull request merged**